compiler/src/dotty/tools/dotc/typer/Namer.scala

```diff
@@ -374,6 +374,10 @@ class Namer { typer: Typer =>
 
         val completer = tree match {
           case tree: TypeDef => new TypeDefCompleter(tree)(cctx)
+          case tree: ValDef if ctx.explicitNulls && ctx.owner.is(Method) && !tree.mods.is(Lazy | Implicit) =>
+            // Use a completer that completes itself with the completion text, so
+            // we can use flow typing for `ValDef`s.
+            new ValDefInBlockCompleter(tree)(cctx)
```

Comment:
Added a comment:
            // Don't use this special kind of completer for `lazy` or `implicit` symbols,
            // because those could be completed through a forward reference:
            // e.g.
            //   val x: String|Null = ???
            //   y /* y is completed through a forward reference! */
            //   if (x == null) throw new NullPointerException()
            //   lazy val y: Int = x.length